### ISSUE TITLE
routes.json just stopped working

### ISSUE THREAD

**Issue by benc-uk** [Azure/static-web-apps#52]
Our project was using routes.json fine, then for no reason they have stopped working.
There's nothing clever in there, just a single route now, but it is never picked up.

https://github.com/benc-uk/taskr/blob/master/public/routes.json

In the build logs I can see it being detected and validated ok
https://github.com/benc-uk/taskr/runs/750508642?check_suite_focus=true#step:4:211

How are routes.json implemented? Via Functions Proxies or some other mechanism?
How can we debug this?
(cc @jasoncabot-ms)

**Comment by benc-uk** [comment]
Tried deleting the static web app in Azure and redeployed it, and the same problem

Has the format of the routes.json file changed? Is anyone else affected?

**Comment by benc-uk** [comment]
And now mysteriously it's sprung back into life
😱

**Comment by benc-uk** [comment]
Seems like 301 redirects work ok but other routes will not, e.g.

```
{
  "routes": [
    // This route will NOT work
    {
      "route": "/test/*",
      "serve": "/api/debugger",
      "statusCode": 200
    },
    // This route work
    {
      "route": "/test301",
      "serve": "/api/debugger",
      "statusCode": 301
    },   
    // This route wil NOT work 
    {
      "route": "/postLogin",
      "serve": "/api/postLogin",
      "statusCode": 200
    },
    // This route wil NOT work 
    {
      "route": "/wibble",
      "serve": "/foo.html",
      "statusCode": 301
    },
    // This route will work for some strange reason
    {
      "route": "/*",
      "serve": "/index.html",
      "statusCode": 200
    }       
  ]
}
```
(Note. I know JSONC probably isn't supported but comments in the JSON was clearest way of illustrating my point)

**Comment by miwebst** [comment]
Hey @benc-uk , thanks for trying this out!

Can you provide more info the behavior you are seeing? Is it giving you a 404?

I would expect this route to work:
`    {
      "route": "/wibble",
      "serve": "/foo.html",
      "statusCode": 301
    },`

I believe we ignore the 'serve' field on redirects. What is the behavior for this route?

**Comment by benc-uk** [comment]
{
  "routes": [
    // This works
    {
      "route": "/test-folder-200",
      "serve": "/test",
      "statusCode": 200
    },

    // This DOES NOT work (404 result)
    {
      "route": "/test-api-200",
      "serve": "/api/debugger",
      "statusCode": 200
    },

    // This works
    {
      "route": "/twitter-login-200",
      "serve": "/.auth/login/twitter",
      "statusCode": 200
    },

    // This DOES NOT work (404 result)
    {
      "route": "/twitter-login-200-query",
      "serve": "/.auth/login/twitter?post_login_redirect_uri=/api/postLogin",
      "statusCode": 200
    },    

    // This works
    {
      "route": "/test-page-200",
      "serve": "/foo.html",
      "statusCode": 200
    },    

    // This works
    {
      "route": "/test-folder-301",
      "serve": "/test",
      "statusCode": 301
    },

    // This works
    {
      "route": "/test-api-301",
      "serve": "/api/debugger",
      "statusCode": 301
    },

    // This works
    {
      "route": "/twitter-login-301",
      "serve": "/.auth/login/twitter",
      "statusCode": 301
    },  

    // This DOES NOT work (404 result)
    {
      "route": "/twitter-login-301-query",
      "serve": "/.auth/login/twitter?post_login_redirect_uri=/api/postLogin",
      "statusCode": 301
    },      

    // This works
    {
      "route": "/test-page-301",
      "serve": "/foo.html",
      "statusCode": 301
    }
  ]
}

**Comment by benc-uk** [comment]
Tried escaping the question mark, e.g.
```
"serve": "/.auth/login/twitter\\?post_login_redirect_uri=/api/postLogin"
```
It still results in a 404

**Comment by mkarmark** [comment]
Hi @benc-uk, 

Our service currently does not support query string parameters appended to the paths assigned to the "serve" variable in the routes file. The values should strictly be paths present in the static web app.

As for the wibble --> foo.html route, I notice that the rule is no longer present in the repo. I would expect that rule to work without a problem, so would it be possible to add that rule back in so I can further investigate why that rule did not work?

Thanks for trying out Azure Static Web Apps and thanks for your patience!

**Comment by mkarmark** [comment]
Regarding the rule 
```
{
      "route": "/test-api-200",
      "serve": "/api/debugger",
      "statusCode": 200
    },
```

Could we please get some context for why you would rather have the path route to the function rather than redirect to the function? We currently don't support routing _to_ functions but we can add support for that if there is an ask.

**Comment by benc-uk** [comment]
I've tried so many things I don't recall if the **wibble --> foo.html** route worked, I suspect that one might have been user error

I think we've got to the bottom of this, it's a case of the two things I was trying to do and testing with, were an exact overlap with two things not supported by the service.

I suggest making [the docs clearer](https://docs.microsoft.com/en-gb/azure/static-web-apps/routes) and highlight two things:
- State the fact that query strings are not supported by the routes file, this is what has cost me the most time. There's no mention of this limitation [here](https://docs.microsoft.com/en-gb/azure/static-web-apps/routes) 
- Mention that routing to functions is also not supported (unless with a redirect), personally I don't have a use case for this (I was simply testing using a backend Function that provided some debug data, header information etc), but I can see many people trying it and being confused why it doesn't work.
